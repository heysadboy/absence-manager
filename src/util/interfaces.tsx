export interface IAbsence {
    admitterId: number | null;
    admitterNote: string | null,
    confirmedAt: string | null,
    createdAt: string | null,
    crewId: number,
    endDate: string | null,
    id: number,
    memberNote: string | null,
    rejectedAt: string | null,
    startDate: string | null,
    type: string | null,
    userId: number
}

export interface IMember {
    crewId: number,
    id: number,
    image: string,
    name: string,
    userId: number
}

export interface IMemberList {
    [id: number]: IMember
}