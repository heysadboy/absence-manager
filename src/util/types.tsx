export type TPageIndex = {
    start: number;
    end: number;
    total: number
}

export type TDateRange = {
    start: string,
    end: string
}