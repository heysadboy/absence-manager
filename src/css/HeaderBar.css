#header-bar {
    padding: 20px;
}

#header-title {
    font-size: x-large;
    font-weight: bold;
}

@media screen and (max-width: 480px) {
    #header-title {
        font-size: medium;
        font-weight: bold;
    }
}