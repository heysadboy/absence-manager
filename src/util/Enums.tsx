export enum EAbsenceStatus {
    requested = 'Requested',
    confirmed = 'Confirmed',
    rejected = 'Rejected'
}

export enum EPageStatus {
    loading = 'Loading',
    error = 'Error',
    success = 'Success',
    empty = 'Empty'
}

export enum EAbsenceType {
    sickness = 'Sickness',
    vacation = 'Vacation',
    all = 'All'
}
