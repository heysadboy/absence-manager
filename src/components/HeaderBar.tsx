import { useState, FC } from 'react';
import { EAbsenceType } from '../util/Enums';
import { TDateRange } from '../util/Types';
import '../css/HeaderBar.css';

interface IHeaderBarProp {
    title: string;
    absenceType: EAbsenceType;
    setDateRange: (dateRange: TDateRange) => void;
    setAbsenceType: (absenceType: EAbsenceType) => void;
}

const HeaderBar: FC<IHeaderBarProp> = ({ title, absenceType, setDateRange, setAbsenceType }) => {
    const [dateError, setDateError] = useState(false);
    const [date, setDate] = useState("");

    const getAbsenceType = (aType: EAbsenceType) => {
        switch (aType) {
            case EAbsenceType.all: return "All absence";
            case EAbsenceType.sickness: return "Sickness";
            case EAbsenceType.vacation: return "Vacation";
        }
    }

    const typeSelect = (e: any) => {
        setAbsenceType(e.target.dataset.value);
    };

    const getPosition = (string: string, subString: string, index: number) => {
        return string.split(subString, index).join(subString).length;
    }

    const getSelectedDate = (e: any) => {
        setDate(e.target.value);
        setDateRange({ start: null, end: null })
        if (date === "") {
            setDateError(false);
        }
    };

    const getDateRange = () => {

        let tempDateRange = date;
        tempDateRange = tempDateRange.replace(/\s/g, '');
        const i = getPosition(tempDateRange, '-', 3);
        const startDate = new Date(tempDateRange.substring(0, i));
        const endDate = new Date(tempDateRange.substring(i + 1));

        if (isNaN(startDate.getTime()) || isNaN(endDate.getTime())) {
            setDateError(true);
        }
        else {
            setDateError(false);
            setDateRange({ start: startDate, end: endDate })
        }
    }

    return (
        <div id="header-bar" className="ui top fixed vertical menu">
            <p id="header-title">{title}</p>
            <div id="date-filter" className={`ui action input ${dateError ? `error` : ``}`}>
                <input value={date} onChange={getSelectedDate} className="ui input" placeholder="yyyy-mm-dd - yyyy-mm-dd" />
                <button onClick={getDateRange} className={`ui ${dateError ? `red` : ``} icon button`}>
                    <i className="calendar icon"></i>
                </button>
            </div>
            <div id="type-filter" className="ui simple selection dropdown">
                <i className="dropdown icon"></i>
                <div className="text">{getAbsenceType(absenceType)}</div>
                <div className="menu">
                    <div className={`${absenceType === EAbsenceType.all ? 'active' : ''} item`} onClick={typeSelect} data-value={EAbsenceType.all}>All absence</div>
                    <div className={`${absenceType === EAbsenceType.sickness ? 'active' : ''} item`} onClick={typeSelect} data-value={EAbsenceType.sickness}>Sickness</div>
                    <div className={`${absenceType === EAbsenceType.vacation ? 'active' : ''} item`} onClick={typeSelect} data-value={EAbsenceType.vacation}>Vacation</div>
                </div>
            </div>
        </div>
    );
};

export default HeaderBar;