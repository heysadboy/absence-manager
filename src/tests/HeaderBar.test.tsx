import renderer from 'react-test-renderer';
import HeaderBar from '../components/HeaderBar';
import { EAbsenceType } from '../util/Enums';

test('HeaderBar.tsx snapshot testing', () => {
    const setDateRange = () => {}
    const setAbsenceType = () => {}

    const component = renderer.create(
        <HeaderBar title="Header Bar" absenceType={EAbsenceType.all} setDateRange={setDateRange} setAbsenceType={setAbsenceType} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});