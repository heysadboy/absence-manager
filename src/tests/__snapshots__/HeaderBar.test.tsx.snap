// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBar.tsx snapshot testing 1`] = `
<div
  className="ui top fixed vertical menu"
  id="header-bar"
>
  <p
    id="header-title"
  >
    Header Bar
  </p>
  <div
    className="ui action input "
    id="date-filter"
  >
    <input
      className="ui input"
      onChange={[Function]}
      placeholder="yyyy-mm-dd - yyyy-mm-dd"
      value=""
    />
    <button
      className="ui  icon button"
      onClick={[Function]}
    >
      <i
        className="calendar icon"
      />
    </button>
  </div>
  <div
    className="ui simple selection dropdown"
    id="type-filter"
  >
    <i
      className="dropdown icon"
    />
    <div
      className="text"
    >
      All absence
    </div>
    <div
      className="menu"
    >
      <div
        className="active item"
        data-value="all"
        onClick={[Function]}
      >
        All absence
      </div>
      <div
        className=" item"
        data-value="sickness"
        onClick={[Function]}
      >
        Sickness
      </div>
      <div
        className=" item"
        data-value="vacation"
        onClick={[Function]}
      >
        Vacation
      </div>
    </div>
  </div>
</div>
`;
